File: 0028. Find the Index of the First Occurrence in a String/0028-Find-the-Index-of-the-First-Occurrence-in-a-String.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int h = 0, n = 0, p = -1;
        while (h < haystack.length()) {
            if (haystack[h] != needle[n]) {
                n = 0;
                if (p != -1) {
                    h = p + 1;
                } else {
                    h = h + 1;
                }
                p = -1;
                continue;
            }

            if (n == 0) {
                p = h;
            }

            h += 1;    
            n += 1;
            
            if (n == needle.length()) {
                return h - n;
            }
        }
        return -1;
    }
};

File: 0028. Find the Index of the First Occurrence in a String/0028-Find-the-Index-of-the-First-Occurrence-in-a-String.java
class Solution {
    public int strStr(String haystack, String needle) {
        int len = needle.length();
        for (int i = 0; i < (haystack.length() - len + 1); i++) {
            if (haystack.substring(i, i + len).equals(needle)) {
                return i;
            }
        }
        return -1;    
    }
}